lib/membrane/rtp/session/report.ex

```diff
@@ -0,0 +1,137 @@
+defmodule Membrane.RTP.Session.Report do
+  @moduledoc false
+  alias Membrane.{RTCP, RTP, Time}
+  require Membrane.Logger
+
+  defmodule Data do
+    @moduledoc false
+    use Bunch.Access
+
+    @type t :: %__MODULE__{
+            remote_ssrcs: MapSet.t(RTP.ssrc_t()),
+            remote_reports: %{
+              RTP.ssrc_t() => %{
+                cut_wallclock_timestamp: pos_integer,
+                arrival_time: Time.t()
+              }
+            },
+            stats: [{RTP.ssrc_t(), RTP.ssrc_t(), RTP.JitterBuffer.Stats.t()}]
+          }
+    defstruct remote_ssrcs: MapSet.new(),
+              remote_reports: %{},
+              stats: []
+  end
+
+  @type maybe_report_t :: {:report, RTCP.CompoundPacket.t()} | :no_report
+
+  @spec init_report(ssrcs :: %{RTP.ssrc_t() => RTP.ssrc_t()}, Data.t()) ::
+          {maybe_report_t, MapSet.t(RTP.ssrc_t()), Data.t()}
+  def init_report(ssrcs, report_data) do
+    remote_ssrcs = ssrcs |> Map.keys() |> MapSet.new()
+
+    maybe_report =
+      if Enum.empty?(report_data.remote_ssrcs) do
+        :no_report
+      else
+        Membrane.Logger.warn(
+          "Not received stats from ssrcs: #{Enum.join(report_data.ssrcs, ", ")}"
+        )
+
+        {:report, generate_report(report_data)}
+      end
+
+    remote_reports =
+      report_data.remote_reports
+      |> Bunch.KVEnum.filter_by_keys(&MapSet.member?(remote_ssrcs, &1))
+      |> Map.new()
+
+    report_data = %{
+      report_data
+      | remote_ssrcs: remote_ssrcs,
+        remote_reports: remote_reports,
+        stats: []
+    }
+
+    {maybe_report, remote_ssrcs, report_data}
+  end
+
+  @spec handle_stats(
+          RTP.JitterBuffer.Stats.t(),
+          RTP.ssrc_t(),
+          %{RTP.ssrc_t() => RTP.ssrc_t()},
+          Data.t()
+        ) ::
+          {maybe_report_t, Data.t()}
+  def handle_stats(stats, remote_ssrc, ssrcs, report_data) do
+    report_ssrcs = MapSet.delete(report_data.remote_ssrcs, remote_ssrc)
+
+    stats =
+      case Map.fetch(ssrcs, remote_ssrc) do
+        {:ok, local_ssrc} -> [{local_ssrc, remote_ssrc, stats}]
+        :error -> []
+      end
+
+    report_data = %{report_data | stats: stats ++ report_data.stats, remote_ssrcs: report_ssrcs}
```

Comment:
Lists concatenation when building them should be avoided. Does the order of stats matter? If you still want to use `++` you can at least swap arguments, so that LHS will always have 1 element at most.

Reply:
well, LHS has one element at most...